lib/internal/fs/copy/copy.js

```diff
@@ -0,0 +1,468 @@
+}
+
+function handleTimestampsAndMode(srcMode, src, dest, cb) {
+  // Make sure the file is writable before setting the timestamp
+  // otherwise open fails with EPERM when invoked with 'r+'
+  // (through utimes call)
+  if (fileIsNotWritable(srcMode)) {
+    return makeFileWritable(dest, srcMode, (err) => {
+      if (err) return cb(err);
+      return setDestTimestampsAndMode(srcMode, src, dest, cb);
+    });
+  }
+  return setDestTimestampsAndMode(srcMode, src, dest, cb);
+}
+
+function fileIsNotWritable(srcMode) {
+  return (srcMode & 0o200) === 0;
+}
+
+function makeFileWritable(dest, srcMode, cb) {
+  return setDestMode(dest, srcMode | 0o200, cb);
+}
+
+function setDestTimestampsAndMode(srcMode, src, dest, cb) {
+  setDestTimestamps(src, dest, (err) => {
+    if (err) return cb(err);
+    return setDestMode(dest, srcMode, cb);
+  });
+}
+
+function setDestMode(dest, srcMode, cb) {
+  return chmod(dest, srcMode, cb);
+}
+
+function setDestTimestamps(src, dest, cb) {
+  // The initial srcStat.atime cannot be trusted
+  // because it is modified by the read(2) system call
+  // (See https://nodejs.org/api/fs.html#fs_stat_time_values)
+  stat(src, (err, updatedSrcStat) => {
+    if (err) return cb(err);
+    return utimesMillis(dest, updatedSrcStat.atime, updatedSrcStat.mtime, cb);
+  });
+}
+
+function utimesMillis(path, atime, mtime, callback) {
+  open(path, 'r+', (err, fd) => {
+    if (err) return callback(err);
+    futimes(fd, atime, mtime, (futimesErr) => {
+      close(fd, (closeErr) => {
+        if (callback) callback(futimesErr || closeErr);
+      });
+    });
+  });
+}
+
+function onDir(srcStat, destStat, src, dest, opts, cb) {
+  if (!destStat) return mkDirAndCopy(srcStat.mode, src, dest, opts, cb);
+  return copyDir(src, dest, opts, cb);
+}
+
+function mkDirAndCopy(srcMode, src, dest, opts, cb) {
+  mkdir(dest, (err) => {
+    if (err) return cb(err);
+    copyDir(src, dest, opts, (err) => {
+      if (err) return cb(err);
+      return setDestMode(dest, srcMode, cb);
+    });
+  });
+}
+
+function copyDir(src, dest, opts, cb) {
+  readdir(src, (err, items) => {
+    if (err) return cb(err);
+    return copyDirItems(items, src, dest, opts, cb);
+  });
+}
+
+function copyDirItems(items, src, dest, opts, cb) {
+  const item = items.pop();
+  if (!item) return cb(null);
+  return copyDirItem(items, item, src, dest, opts, cb);
+}
+
+function copyDirItem(items, item, src, dest, opts, cb) {
+  const srcItem = join(src, item);
+  const destItem = join(dest, item);
+  checkPaths(srcItem, destItem, opts, (err, stats) => {
+    if (err) return cb(err);
+    const { destStat } = stats;
+    startCopy(destStat, srcItem, destItem, opts, (err) => {
+      if (err) return cb(err);
+      return copyDirItems(items, src, dest, opts, cb);
+    });
+  });
+}
+
+function onLink(destStat, src, dest, opts, cb) {
+  readlink(src, (err, resolvedSrc) => {
+    if (err) return cb(err);
+    // TODO(bcoe): I don't know how this could be called, because
```

Comment:
@manidlou @RyanZim, I'm not seeing how we could hit this logic and have `opts.dereference` `true`, because `getStatsForCopy` will have used `stat` and followed the symlink to its src.

I'm probably missing something.

Reply:
This logic was also not exercised by any of the tests in `fs-extra`, I believe I was right that it can't be hit, so removed it for the time being.